Retain generic ARM64 support without Apple CPU MPIDR helpers in the lab.
Match the KPERF field guard, retain native panic and four formatting fragments.
No BSD process context is claimed by firmware panic diagnostics.

diff --git a/osfmk/arm64/pcb.c b/osfmk/arm64/pcb.c
--- a/osfmk/arm64/pcb.c
+++ b/osfmk/arm64/pcb.c
@@ -63,7 +63,9 @@
 
 #include <IOKit/IOBSD.h>
 
+#if !defined(ARM64_BOARD_CONFIG_ATLANTIC_LAB)
 #include <pexpert/arm64/apple_arm64_cpu.h>
+#endif
 #include <pexpert/pexpert.h>
 
 // fixme: rdar://114299113 tracks resolving the supportlib issue with hwtrace features
diff --git a/osfmk/arm/machine_routines_common.c b/osfmk/arm/machine_routines_common.c
--- a/osfmk/arm/machine_routines_common.c
+++ b/osfmk/arm/machine_routines_common.c
@@ -49,7 +49,9 @@
 #include <machine/monotonic.h>
 #include <pexpert/pexpert.h>
 #include <pexpert/device_tree.h>
+#if !defined(ARM64_BOARD_CONFIG_ATLANTIC_LAB)
 #include <pexpert/arm64/apple_arm64_cpu.h>
+#endif
 
 #include <mach/machine.h>
 #include <mach/machine/sdt.h>
diff --git a/osfmk/kern/ast.c b/osfmk/kern/ast.c
--- a/osfmk/kern/ast.c
+++ b/osfmk/kern/ast.c
@@ -262,7 +262,9 @@
 		kpc_thread_ast_handler(thread);
 #endif /* CONFIG_CPU_COUNTERS */
 		kperf_thread_ast_handler(thread);
+#if CONFIG_KPERF
 		thread->kperf_ast = 0;
+#endif
 	}
 
 	if (reasons & AST_RESET_PCS) {
diff --git a/osfmk/kern/debug.c b/osfmk/kern/debug.c
--- a/osfmk/kern/debug.c
+++ b/osfmk/kern/debug.c
@@ -88,6 +88,9 @@
 #include <console/serial_protos.h>
 #include <IOKit/IOBSD.h>
 #include <libkern/crc.h>
+#if CONFIG_ATLANTIC_FIRMWARE
+#include <kern/atlantic_printf.h>
+#endif
 
 #if !(MACH_KDP && CONFIG_KDP_INTERACTIVE_DEBUGGING)
 #include <kdp/kdp_udp.h>
@@ -1169,12 +1172,19 @@
 		*task = thread->t_tro->tro_task;
 	}
 
+#if CONFIG_ATLANTIC_FIRMWARE && !MACH_BSD
+	/* No BSD process context is available in this firmware profile. */
+	if (proc) {
+		*proc = NULL;
+	}
+#else
 	if (!panic_validate_ptr(thread->t_tro->tro_proc,
 	    sizeof(struct proc *), "bsd_info")) {
 		*proc = NULL;
 	} else {
 		*proc = thread->t_tro->tro_proc;
 	}
+#endif
 
 	return true;
 }
diff --git a/osfmk/kern/atlantic_printf.c b/osfmk/kern/atlantic_printf.c
new file mode 100644
--- /dev/null
+++ b/osfmk/kern/atlantic_printf.c
@@ -0,0 +1,115 @@
+/*
+ * Copyright (c) 2000-2006 Apple Computer, Inc. All rights reserved.
+ *
+ * @APPLE_OSREFERENCE_LICENSE_HEADER_START@
+ *
+ * This file contains Original Code and/or Modifications of Original Code
+ * as defined in and that are subject to the Apple Public Source License
+ * Version 2.0 (the 'License'). You may not use this file except in
+ * compliance with the License. The rights granted to you under the License
+ * may not be used to create, or enable the creation or redistribution of,
+ * unlawful or unlicensed copies of an Apple operating system, or to
+ * circumvent, violate, or enable the circumvention or violation of, any
+ * terms of an Apple operating system software license agreement.
+ *
+ * Please obtain a copy of the License at
+ * http://www.opensource.apple.com/apsl/ and read it before using this file.
+ *
+ * The Original Code and all software distributed under the License are
+ * distributed on an 'AS IS' basis, WITHOUT WARRANTY OF ANY KIND, EITHER
+ * EXPRESS OR IMPLIED, AND APPLE HEREBY DISCLAIMS ALL SUCH WARRANTIES,
+ * INCLUDING WITHOUT LIMITATION, ANY WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE, QUIET ENJOYMENT OR NON-INFRINGEMENT.
+ * Please see the License for the specific language governing rights and
+ * limitations under the License.
+ *
+ * @APPLE_OSREFERENCE_LICENSE_HEADER_END@
+ */
+/* Copyright (c) 1995 NeXT Computer, Inc. All Rights Reserved */
+/*-
+ * Copyright (c) 1986, 1988, 1991, 1993
+ *	The Regents of the University of California.  All rights reserved.
+ * (c) UNIX System Laboratories, Inc.
+ * All or some portions of this file are derived from material licensed
+ * to the University of California by American Telephone and Telegraph
+ * Co. or Unix System Laboratories, Inc. and are reproduced herein with
+ * the permission of UNIX System Laboratories, Inc.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ * 3. All advertising materials mentioning features or use of this software
+ *    must display the following acknowledgement:
+ *	This product includes software developed by the University of
+ *	California, Berkeley and its contributors.
+ * 4. Neither the name of the University nor the names of its contributors
+ *    may be used to endorse or promote products derived from this software
+ *    without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE REGENTS AND CONTRIBUTORS ``AS IS'' AND
+ * ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED.  IN NO EVENT SHALL THE REGENTS OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+ * OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+ * HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF
+ * SUCH DAMAGE.
+ *
+ *	@(#)subr_prf.c	8.4 (Berkeley) 5/4/95
+ */
+/* Native formatting subset retained verbatim for the Mach-only firmware. */
+#include <kern/atlantic_printf.h>
+#include <kern/misc_protos.h>
+#include <mach/boolean.h>
+
+struct snprintf_arg {
+	char *str;
+	size_t remain;
+};
+
+static void
+snprintf_func(int ch, void *arg)
+{
+	struct snprintf_arg *const info = arg;
+
+	if (info->remain >= 2) {
+		*info->str++ = (char)ch;
+		info->remain--;
+	}
+}
+
+int
+vsnprintf(char *str, size_t size, const char *format, va_list ap)
+{
+	struct snprintf_arg info;
+	int retval;
+
+	info.str = str;
+	info.remain = size;
+	retval = __doprnt(format, ap, snprintf_func, &info, 10, FALSE);
+	if (info.remain >= 1) {
+		*info.str++ = '\0';
+	}
+	return retval;
+}
+
+int
+snprintf(char *str, size_t size, const char *format, ...)
+{
+	int retval;
+	va_list ap;
+
+	va_start(ap, format);
+	retval = vsnprintf(str, size, format, ap);
+	va_end(ap);
+	return retval;
+}
+
diff --git a/osfmk/kern/atlantic_printf.h b/osfmk/kern/atlantic_printf.h
new file mode 100644
--- /dev/null
+++ b/osfmk/kern/atlantic_printf.h
@@ -0,0 +1,9 @@
+/* SPDX-License-Identifier: MIT */
+/* AtlanticOS declarations for retained native XNU formatting routines. */
+#ifndef ATLANTIC_PRINTF_H
+#define ATLANTIC_PRINTF_H
+#include <stddef.h>
+#include <stdarg.h>
+extern int snprintf(char *, size_t, const char *, ...) __attribute__((format(printf, 3, 4)));
+extern int vsnprintf(char *, size_t, const char *, va_list) __attribute__((format(printf, 3, 0)));
+#endif
